pyccel/parser/semantic.py

```diff
@@ -1430,6 +1459,24 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign, *
                             bounding_box=(self._current_fst_node.lineno, self._current_fst_node.col_offset),
                             severity='fatal', blocker=False)
 
+                elif var.is_ndarray and isinstance(rhs, (Variable, IndexedElement)) and var.allocatable:
+                    errors.report(ASSIGN_ARRAYS_ONE_ANOTHER,
+                        bounding_box=(self._current_fst_node.lineno,
+                            self._current_fst_node.col_offset),
+                                severity='error', symbol=lhs.name)
+
+                elif var.is_ndarray and var.is_target:
+                    errors.report(ARRAY_ALREADY_IN_USE,
+                        bounding_box=(self._current_fst_node.lineno,
+                            self._current_fst_node.col_offset),
+                                severity='error', symbol=var.name)
+
+                elif var.is_ndarray and var.is_pointer:
+                    # we allow pointers to be reassigned multiple times
+                    # pointers reassigning need to call free_pointer func
+                    # to remove memory leaks
+                    new_expressions.append(Deallocate(var))
```

Comment:
I wonder if we should find a way to differentiate between deallocating memory and deallocating shapes. In fortran we can call `allocate` on both `allocatable` variables and `pointer` variables. The only differences are that `allocatable` must be contiguous, and `allocatable` is automatically deallocated at the end of the function/subroutine. We will therefore need to allocate `pointer` variables for issue #537. once this is done the current implementation of `Deallocate` might prevent us from writing a `_print_Deallocate` for fortran.

What do you think?

Reply:
the problem with pointers is that we can't know if they own their data or not statically

Reply:
we can have call `_print_Deallocate` only for pointers but there is no way to know whats the pointer is holding statically 
- because we are not tracking the changes that occurs to a pointer variables (don't know if this is a bug)
- we can't determine the outcome of an `if block`